README.md
# slab-PDC

## Copyright

Authors: Paolo Missiaggia, <paolo.ratm@gmail.com>

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License version 2 as published by
the Free Software Foundation.

Copyright (C) 2016-2017 - Paolo Missiaggia - All Rights Reserved

## About

SLAB-PDC, a memory allocator with pre-distruction callback (PDC).

Allow to call a pre-distructor function before actually freeing memory.

## Tech info

Requirements:
The first 2 element of the returned struct SHALL be:
* "struct rcu_head"
* "pdc_func_f"

E.g.:

```C
struct foo {
   struct rcu_head rcu;
   pdc_func_f pdc_func;
   int foo1;
};
```
or:

```C
struct foo {
   pdc_t kobj;
   int foo1;
};
```

AAA: Do not perform non atomic actions inside "pdc_func"
AAA: The pdc_func SHALL not free the memory itself


ChangeLog
[slab_pdc 1.1.8] Changelog

* [MK] fixed issue with superuser base.sh

[slab_pdc 1.1.7] Changelog

* [MK] Added build scripts to tools

[slab_pdc 1.1.6] Changelog

* Update Make install commands to generate and install DKMS deb

[slab_pdc 1.1.5] Changelog

* Add debber.sh to fulfill the build machine requirements

[slab_pdc 1.1.4] Changelog

* Export symbols file in a version agnostic path

[slab_pdc 1.1.3] Changelog

* Add patch field to versioning

[slab_pdc 1.1.2] Changelog

* Add DKMS template file to avoid DEB creation problems

[slab_pdc 1.1.1] Changelog

* Fix DEB creation to correctly interact with depmod in post_inst
* Update file headers with a properly formed template
* Add legal notices to the README file

[slab_pdc 1.1.0] Changelog

* Remove verison header
* Add Makefile rule to create DKMS deb packet

[slab_pdc 1.1.0] Changelog

*slab PDC - A first implementation and sample


slab_pdc.c
/*
 * SLAB-PDC
 * A memory allocator with pre-distruction callback (PDC).
 * Copyright (C) 2016-2017 - Paolo Missiaggia - All Rights Reserved
 *
 * Authors:
 *         Paolo Missiaggia, <paolo.ratm@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 */

#include <linux/module.h>
#include <linux/kernel.h>
#include <linux/moduleparam.h>
#include <linux/version.h>
#include <linux/types.h>
#include <linux/rcupdate.h>

#include "slab_pdc.h"

__always_inline static void
call_rcu_pdc(struct rcu_head *head)
{
	pdc_func_f pdc_func;
	pdc_t *p = NULL;

	p = container_of(head, pdc_t, rcu);
	if (!p)
		return;

	pdc_func = p->pdc_func;
	if (pdc_func)
		pdc_func(p);

	kfree(p);
}

void *
kzalloc_pdc(size_t size, gfp_t flags, pdc_func_f pdc_func)
{
	pdc_t *p = NULL;

	p = kzalloc(size, flags);
	if (unlikely(!p))
		return (NULL);

	p->pdc_func = pdc_func;

	return (p);
}
EXPORT_SYMBOL(kzalloc_pdc);

void
kfree_pdc(const void *obj)
{
	pdc_t *p = NULL;

	if (!obj)
		return;

	p = (pdc_t *)obj;
	if (p->pdc_func)
		call_rcu(&p->rcu, call_rcu_pdc);
	else
		kfree_rcu(p, rcu);
}
EXPORT_SYMBOL(kfree_pdc);

void
pdc_wait_on_exit(void)
{
	return(rcu_barrier());
}
EXPORT_SYMBOL(pdc_wait_on_exit);

int
slab_pdc_init(void)
{
	printk(KERN_INFO "[%s] module loaded\n", SLAB_PDC_VERSION_STR);

	return 0;
}

void
slab_pdc_exit(void)
{
	pdc_wait_on_exit();

	printk(KERN_INFO "[%s] module unloaded\n", SLAB_PDC_VERSION_STR);
}

module_init(slab_pdc_init);
module_exit(slab_pdc_exit);

MODULE_AUTHOR("Paolo Missiaggia <paolo.ratm@gmail.com>");
MODULE_DESCRIPTION("Kernel memory handling facilities");
MODULE_VERSION(SLAB_PDC_VERSION_STR);

MODULE_LICENSE("GPL");


slab_pdc_version.h
/*
 * Memory allocator with pre-distruction callback (PDC)
 * (C) 2016
 *
 * Authors:
 *         Paolo Missiaggia, <paolo.ratm@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 */

#ifndef SLAB_PDC_VERSION_H
#define SLAB_PDC_VERSION_H

#define SLAB_PDC_STR_H(x) #x
#define SLAB_PDC_STR(x) SLAB_PDC_STR_H(x)

#define SLAB_PDC_MAINTAINER "paolo.ratm@gmail.com"

#define SLAB_PDC_NAME  "SLAB_PDC"
#define SLAB_PDC_MAJOR 1
#define SLAB_PDC_MINOR 0
#define SLAB_PDC_PATCH 0

#define SLAB_PDC_VERSION(a, b, c) (((a) << 24) + ((b) << 16) + (c))

#define SLAB_PDC_VERSION_CODE \
	SLAB_PDC_VERSION(SLAB_PDC_MAJOR, SLAB_PDC_MINOR, SLAB_PDC_PATCH)

/* #define SLAB_PDC_ALPHA 1 */
/* #define SLAB_PDC_BETA 1 */

#if defined(SLAB_PDC_ALPHA)
#define SLAB_PDC_VERSION_STR \
	SLAB_PDC_NAME ": "       \
	SLAB_PDC_STR(SLAB_PDC_MAJOR)  \
	"."                 \
	SLAB_PDC_STR(SLAB_PDC_MINOR)  \
	"."                 \
	SLAB_PDC_STR(SLAB_PDC_PATCH)  \
	"-A"                \
	SLAB_PDC_STR(SLAB_PDC_ALPHA)
#elif defined(SLAB_PDC_BETA)
#define SLAB_PDC_VERSION_STR \
	SLAB_PDC_NAME ": "       \
	SLAB_PDC_STR(SLAB_PDC_MAJOR)  \
	"."                 \
	SLAB_PDC_STR(SLAB_PDC_MINOR)  \
	"."                 \
	SLAB_PDC_STR(SLAB_PDC_PATCH)  \
	"-B"                \
	SLAB_PDC_STR(SLAB_PDC_BETA)
#else
#define SLAB_PDC_VERSION_STR \
	SLAB_PDC_NAME ": "       \
	SLAB_PDC_STR(SLAB_PDC_MAJOR)  \
	"."                 \
	SLAB_PDC_STR(SLAB_PDC_MINOR)  \
	"."                 \
	SLAB_PDC_STR(SLAB_PDC_PATCH)
#endif


/*
 * E.g.
 *
 * #if SLAB_PDC_VERSION_CODE < SLAB_PDC_VERSION(0, 1, 0)
 *	...
 *	printk(SLAB_PDC_VERSION_STR);
 *	...
 * #else
 *	...
 * #endif
 */

#endif

scripts/make_tarball.sh
#!/bin/bash

## simple check to make sure we're in the right directory....
#git log &> /dev/null
#
#if [ $? -ne 0 ]; then
#	echo "Don't appear to be in the git directory."
#	exit 1
#fi
#
#RPM_SOURCES=$(rpm --eval '%{_sourcedir}' 2> /dev/null)
#
#if [ ! -d "$RPM_SOURCES" ]; then
#	echo "I want to put the tarball into the RPM_SOURCES sources directory, but \`rpm --eval '%{_sourcedir}'\` didn't return a directory that exists."
#	exit 1
#fi
#
#git_changes_work=$(git diff 2> /dev/null | wc -l)
#git_changes_index=$(git diff --cached 2> /dev/null | wc -l)
#
## add the two wc together
#git_changes=$(expr "$git_changes_work" "+" "$git_changes_index")
#
#if [ $git_changes -gt 0 ]; then
#
#	echo "You have differences between your working tree and/or index from the current HEAD."
#	echo "Please commit them before building the tarball."
#
#	exit 1
#
#fi

vfile=slab_pdc_version.h

NAME=`awk '/#define/{ print $2 " " $3}' ${vfile} | awk '/_NAME/{gsub("\"","");print $2}' | awk '{print tolower($0)}'`
MAJOR=`awk '/#define/{ print $2 " " $3}' ${vfile} | awk '/_MAJOR/{print $2}'`
MINOR=`awk '/#define/{ print $2 " " $3}' ${vfile} | awk '/_MINOR/{print $2}'`
PATCH=`awk '/#define/{ print $2 " " $3}' ${vfile} | awk '/_PATCH/{print $2}'`
MAINTAINER=`awk '/#define/{ print $2 " " $3}' ${vfile} | awk '/_MAINTAINER/{gsub("\"","");print $2}' | awk '{print tolower($0)}'`

version=$MAJOR.$MINOR.$PATCH
#version=$(grep VERSION module.c | cut -d '"' -f 2)

make clean

BRANCH=$(git branch | grep '^\*' | awk '{print $2}')

echo "Making tarball from the \"$BRANCH\" branch => ./slab_pdc-$version.tar.gz"
#echo "Making tarball from the \"$BRANCH\" branch => $RPM_SOURCES/slab_pdc-$version.tar.gz"

git archive --format=tar --prefix="slab_pdc-$version/" $BRANCH | gzip -9 > "./slab_pdc-$version.tar.gz"
#git archive --format=tar --prefix="slab_pdc-$version/" $BRANCH | gzip -9 > "$RPM_SOURCES/slab_pdc-$version.tar.gz"



slab_pdc.h
/*
 * SLAB-PDC
 * A memory allocator with pre-distruction callback (PDC).
 * Copyright (C) 2016-2017 - Paolo Missiaggia - All Rights Reserved
 *
 * Authors:
 *         Paolo Missiaggia, <paolo.ratm@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 */

#ifndef SLAB_PDC_H
#define SLAB_PDC_H

#include <linux/slab.h>

#define SLAB_PDC_NAME  "SLAB_PDC"
#define SLAB_PDC_MAJOR 1
#define SLAB_PDC_MINOR 1
#define SLAB_PDC_PATCH 8
#define SLAB_PDC_STR_H(x) #x
#define SLAB_PDC_STR(x) SLAB_PDC_STR_H(x)
#define SLAB_PDC_VERSION(a, b, c) (((a) << 24) + ((b) << 16) + (c))
#define SLAB_PDC_VERSION_STR \
	SLAB_PDC_NAME ": " SLAB_PDC_STR(SLAB_PDC_MAJOR) "." \
	SLAB_PDC_STR(SLAB_PDC_MINOR) "." SLAB_PDC_STR(SLAB_PDC_PATCH)

/*
 * Requirements:
 * The first 2 element of the returned struct SHALL be:
 * 1) a "struct rcu_head"
 * 2) a "pdc_func_f"
 *
 * E.g.:
 * struct foo {
 *    struct rcu_head rcu;
 *    pdc_func_f pdc_func;
 *    int foo1;
 * };
 * or:
 * struct foo {
 *    pdc_t kobj;
 *    int foo1;
 * };
 *
 * AAA: Do not perform non atomic actions inside "pdc_func"
 * AAA: dst_func SHALL not free the memory itself
 */

typedef int(*pdc_func_f)(void *obj);

typedef struct {
	struct rcu_head rcu;
	pdc_func_f pdc_func;
} pdc_t;

void *kzalloc_pdc(size_t size, gfp_t flags, pdc_func_f pdc_func);
void  kfree_pdc(const void *obj);
void  pdc_wait_on_exit(void);

#endif
